apps/settings/js/modules/addon_manager.js

```diff
@@ -0,0 +1,201 @@
+      navigator.mozApps.mgmt.setEnabled(addon, false);
+    }
+  }
+
+  function canDelete(addon) {
+    return isAddon(addon) && addon.removable;
+  }
+
+  function deleteAddon(addon) {
+    if (!isAddon(addon)) {
+      return Promise.reject('not an addon');
+    }
+
+    if (!addon.removable) {
+      return Promise.reject('addon is not deletable');
+    }
+
+    return new Promise(function(resolve, reject) {
+      var request = navigator.mozApps.mgmt.uninstall(addon);
+      request.onsuccess = function() {
+        resolve();
+      };
+      request.onerror = function() {
+        reject();
+      };
+    });
+  }
+
+  function shareAddon(addon) {
+    if (!isAddon(addon)) {
+      return Promise.reject('not an addon');
+    }
+
+    return new Promise(function(resolve, reject) {
+      var activity = new MozActivity({
+        name: 'share',
+        data: {
+          type: 'app',
+          app: addon.origin
+        }
+      });
+
+      activity.onsuccess = function() {
+        resolve();
+      };
+      activity.onerror = function() {
+        reject(activity.error.name);
+      };
+    });
+  }
+
+  // Addons can only apply to apps with the same or lower privilege
+  function privilegeCheck(addon, app) {
+    var addonManifest = getManifest(addon);
+    var appManifest = getManifest(app);
+    return addonManifest.type === appManifest.type ||
+      addonManifest.type === 'certified' ||
+      (addonManifest.type === 'privileged' && appManifest.type !== 'certified');
+  }
+
+  //
+  // We need to keep this code in sync with the matching gecko
+  // code in dom/apps/UserCustomizations.jsm so that our list of
+  // targeted apps actually matches the apps that Gecko injects
+  // the addon into.
+  //
+  function getAddonTargets(addon) {
+    if (!isAddon(addon)) {
+      return Promise.reject('not an addon');
+    }
+
+    var manifest = getManifest(addon);
+    var customizations = manifest && manifest.customizations;
+
+    // If the addon does not specify any customizations, then we know
+    // that it does not target any apps
+    if (!customizations || customizations.length === 0) {
+      return Promise.resolve([]);
+    }
+
+    return AppsCache.apps().then(function(apps) {
+
+      var targets = new Set();
+
+      customizations.forEach(function(customization) {
+        var filterText = customization && customization.filter;
+        if (!filterText) {
+          return;
+        }
+        var filter = new RegExp(filterText);
+
+        apps.forEach(function(app) {
+          if (!getManifest(app).role &&    // ignore addons, themes, etc.
+              filter.test(app.origin) &&   // app origin matches addon filter
+              privilegeCheck(addon, app))  // addon privilege >= than app
+          {
+            targets.add(app);
+          }
+        });
+      });
+
+      // Now convert the set to an array and return it
+      var targetArray = [];
+      targets.forEach((target) => targetArray.push(target));
```

Comment:
Because the Set() removes duplicates. If an addon has multiple customizations with overlapping filters, the same app can be matched more than once.

Now that you mention it, though, if the outer loop was apps and the inner loop was through all the filters, then I could break out of the inner loop on the first match.